src/chat.cpp

```diff
@@ -1131,6 +1225,152 @@ CCMD( say_team )
+			{
+				Printf( "There's no other player to send a private message to.\n" );
+				return false;
+			}
+		}
+		while (( PLAYER_IsValidPlayer( g_ulChatPlayer ) == false ) || ( g_ulChatPlayer == (ULONG)consoleplayer ));
+	}
+
+	return true;
+}
+
+void chat_PrivateMessage( FCommandLine &argv, const ULONG ulPlayer )
+{
+	ULONG		ulIdx;
+	FString		ChatString;
+
+	// [AK] Mods are not allowed to say anything in the player's name.
+	if ( ACS_IsCalledFromConsoleCommand( ) )
+		return;
+
+	// [AK] No chatting while playing a demo.
+	if ( CLIENTDEMO_IsPlaying( ) == true )
+	{
+		Printf( "You can't send private messages during demo playback.\n" );
+		return;
+	}
+
+	if ( argv.argc( ) >= 2 )
+	{
+		// [AK] Don't send private messages to invalid players.
+		if ( ulPlayer == MAXPLAYERS + 1 )
+		{
+			Printf( "There isn't a player named %s" TEXTCOLOR_NORMAL ".\n", argv[1] );
+			return;
+		}
+
+		// [AK] Don't send private messages to ourselves.
+		if (( ulPlayer == (ULONG)consoleplayer ) && ( NETWORK_GetState( ) != NETSTATE_SERVER ))
+		{
+			Printf( "You can't send a private message to yourself.\n" );
+			return;
+		}
+
+		g_ulChatPlayer = ulPlayer;
+
+		if ( argv.argc( ) > 2 )
+		{
+			for ( ulIdx = 2; ulIdx < static_cast<unsigned int>( argv.argc( ) ); ulIdx++ )
+				ChatString.AppendFormat( "%s ", argv[ulIdx] );
+
+			// Send the server's chat string out to clients, and print it in the console.
+			if ( NETWORK_GetState( ) == NETSTATE_SERVER )
+				SERVER_SendChatMessage( MAXPLAYERS, CHATMODE_PRIVATE_SEND, ChatString.GetChars( ), g_ulChatPlayer );
+			else
+				// We typed out our message in the console or with a macro. Go ahead and send the message now.
+				chat_SendMessage( CHATMODE_PRIVATE_SEND, ChatString.GetChars( ) );
+
+			return;
+		}
+	}
+	
+	if (( NETWORK_GetState( ) != NETSTATE_SERVER ) && ( chat_CanSendPrivateMessage( ulPlayer ) ))
+	{
+		// The message we send will only be for the player who we wish to chat with.
+		chat_SetChatMode( CHATMODE_PRIVATE_SEND );
+
+		// Hide the console.
+		C_HideConsole( );
+
+		// Clear out the chat buffer.
+		g_ChatBuffer.Clear( );
+	}
+}
+
+CCMD( sayto )
+{
+	ULONG ulPlayer = 0;
+
+	if ( argv.argc() >= 2 )
+	{
+		if ( stricmp( argv[1], "Server" ) == 0 )
+		{
+			if ( NETWORK_GetState( ) == NETSTATE_SERVER )
+			{
+				Printf( "You can't send a private message to yourself.\n" );
+				return;
+			}
+			else if ( NETWORK_GetState( ) != NETSTATE_CLIENT )
+			{
+				Printf( "You can't send a private message to the server in an offline game.\n" );
+				return;
+			}
+
+			ulPlayer = MAXPLAYERS;
+		}
+		else
+		{
+			ulPlayer = SERVER_GetPlayerIndexFromName( argv[1], true, true );
+			if ( ulPlayer == MAXPLAYERS ) ulPlayer++;
+		}
+	}
+
+	chat_PrivateMessage( argv, ulPlayer );
+}
+
+CCMD( sayto_idx )
+{
+	int playerIndex = 0;
+
+	if ( argv.argc() >= 2 )
+	{
+		if ( argv.SafeGetNumber( 1, playerIndex ) == false )
+			return;
+
+		if ( PLAYER_IsValidPlayer( playerIndex ) == false )
```

Comment:
Should be allow the server to be contacted with idx "-1"?

Reply:
Yes, this should be allowed too.